### PR TITLE
Improve support for projects created outside of VS Code

### LINKED ISSUE #203: Better support for exsting Functions apps
There is some degradation in the experience for users that have existing Functions Apps when it comes to locally debugging the app. Here are a few ideas for improving that.

Note that `$ func init` creates a launch.json that doesn't include starting up the functions host. I can work with them to get the tasks.json added that matches ours, that may help a bit.

### Contribute Functions launch config to the debugger

There is a way to contribute generic launch configs to the debugger that show up if a user hits F5 and VS Code doesn't know how to launch. 

image

### Allow user to setup launch config

I know you can reinitialize the app to write out the launch config, but it's not super discoverable for users. Maybe add a button to "setup Functions debugging" that will add/overwrite the launch config and tasks.
 
### Better handling of language and runtime selection

The prompts for understanding the workspace ask for the runtime twice. Repro:

```sh
$ func init
$ code .
```

Now add a new function to the app, you'll be promoted to select the runtime version, set the runtime. Next you get prompted for the language, then again for the runtime version.

It seems like the first check could be skipped completely.

### PR DESCRIPTION
**Pull request by EricJizbaMSFT**
On extension activation, we will check for un-initialized functions projects based on the following criteria:
1. host.json and local.settings.json exist at the root of the folder
1. projectLanguage and projectRuntime are not set

Then we will display this message:
screen shot 2018-02-26 at 3 06 41 pm

If they select yes, we will initialize the project in the following manner:
1. If we can uniquely detect the language - use that. If we can't, prompt the user for the langauge
1. If we can detect the installed runtime - use that. Otherwise, prompt the user for the runtime
1. Remove '.vscode' from the gitignore if applicable

Related to that work, this PR also includes the following:
1. A command to "Initialize for use with VS Code" that can be run from the command palette
1. A new docs page that will be used with the 'See more info' aka.ms link
1. More tests 😊

Fixes #203

### CODE REVIEW

File: src/commands/createNewProject/CSharpProjectCreator.ts [hunk 1]
```diff
@@ -57,11 +54,11 @@ export class CSharpProjectCreator implements IProjectCreator {
             throw new Error(localize('unrecognizedTargetFramework', 'Unrecognized target framework in project file "{0}".', csProjName));
         } else {
             const targetFramework: string = matches[1];
```
Comment (nturinski) [inline src/commands/createNewProject/CSharpProjectCreator.ts]: Why matches[1] and not [0] out of curiosity?

File: src/commands/createNewProject/CSharpProjectCreator.ts [hunk 2]
```diff
@@ -129,32 +128,34 @@ export class CSharpProjectCreator implements IProjectCreator {
...
-            if (lineMatches !== null && lineMatches.length > 0) {
-                const line: string = lineMatches[0];
-                const versionMatches: RegExpMatchArray | null = /Version=(?:"([^"]+)"|'([^']+)')/g.exec(line);
-                if (versionMatches !== null && versionMatches.length > 2) {
-                    const version: SemVer = new SemVer(versionMatches[1] || versionMatches[2]);
-                    this._telemetryProperties.cSharpFuncSdkVersion = version.raw;
-                    if (version.compare(minVersion) < 0) {
-                        const newContents: string = csprojContents.replace(line, line.replace(version.raw, minVersion));
-                        await fse.writeFile(csprojPath, newContents);
+        if (!isWindows) {
```
Comment (nturinski) [inline src/commands/createNewProject/CSharpProjectCreator.ts]: Does this function fail silently if it's a Windows OS or is this validation not necessary?

File: src/commands/createNewProject/IProjectCreator.ts [hunk 3]
```diff
@@ -3,21 +3,46 @@
...
 import { localize } from "../../localize";
-import { extensionPrefix, ProjectRuntime, TemplateFilter } from "../../ProjectSettings";
+import { extensionPrefix, ProjectRuntime, promptForProjectRuntime, TemplateFilter } from "../../ProjectSettings";
+import { functionRuntimeUtils } from "../../utils/functionRuntimeUtils";
 
-export interface IProjectCreator {
-    deploySubpath?: string;
-    runtime: ProjectRuntime;
-    templateFilter: TemplateFilter;
+export abstract class ProjectCreatorBase {
```
Comment (nturinski) [inline src/commands/createNewProject/IProjectCreator.ts]: Should change the name of the file as this is no longer an interface

File: src/commands/createNewProject/IProjectCreator.ts [hunk 4]
```diff
@@ -3,21 +3,46 @@
...
+        this.telemetryProperties = telemetryProperties;
+    }
+
+    public async getRuntime(): Promise<ProjectRuntime> {
+        // tslint:disable-next-line:strict-boolean-expressions
+        return await functionRuntimeUtils.tryGetLocalRuntimeVersion() || await promptForProjectRuntime(this.ui);
+    }
+
+    public getLaunchJson(): {} | undefined {
+        return undefined;
```
Comment (nturinski) [inline src/commands/createNewProject/IProjectCreator.ts]: Is this always supposed to default to undefined?

File: src/commands/createNewProject/CSharpProjectCreator.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply src/commands/createNewProject/CSharpProjectCreator.ts]: matches[0] is always the _full_ match and the rest of the matches in a RegExpMatchArray are from the capturing groups. So this case:
```
console.log(matches[0]); // Prints "<TargetFramework>net461<\/TargetFramework>"
console.log(matches[1]); // Prints "net461"
```

File: src/commands/createNewProject/CSharpProjectCreator.ts (hunk 2 above)
Comment (EricJizbaMSFT) [reply src/commands/createNewProject/CSharpProjectCreator.ts]: Validation isn't necessary on Windows - 1.0.8 is only required for Mac and Linux

File: src/commands/createNewProject/IProjectCreator.ts (hunk 3 above)
Comment (EricJizbaMSFT) [reply src/commands/createNewProject/IProjectCreator.ts]: Sounds good. Just filed an issue to address separately: https://github.com/Microsoft/vscode-azurefunctions/issues/251

This isn't the only file I want to rename and I'd prefer to do it separately

File: src/commands/createNewProject/IProjectCreator.ts (hunk 4 above)
Comment (EricJizbaMSFT) [reply src/commands/createNewProject/IProjectCreator.ts]: Yep - only like 3 out of 10-ish languages actually support a launch.json (granted those 3 languages are the 3 we care about most - C#, Java, and JavaScript. The rest are just preview langagues)

**Pull request merged**